Milestone 2 update (#90)
matter-labs/zksync-web-v2-docs: docs/dev/guide/hello-world.md

  console.log(`Running deploy script for the Greeter contract`);

  // Initialize the wallet.
  const provider = new Provider(hre.userConfig.zkSyncDeploy?.zkSyncNetwork);
  const wallet = new Wallet("<WALLET-PRIVATE-KEY>");
  
  // Create deployer object and load the artifact of the contract we want to deploy.
  const deployer = new Deployer(hre, wallet);
  const artifact = await deployer.loadArtifact("Greeter");

  // Estimate contract deployment fee
  const greeting = "Hi there!";
  const deploymentFee = await deployer.estimateDeployFee(artifact, [greeting]);

  // Deposit funds to L2
  const depositHandle = await deployer.zkWallet.deposit({
    to: deployer.zkWallet.address,
    token: utils.ETH_ADDRESS,
    amount: deploymentFee.mul(2),
  });
  // Wait until the deposit is processed on zkSync
  await depositHandle.wait();

  // Deploy this contract. The returned object will be of a `Contract` type, similarly to ones in `ethers`.
  // `greeting` is an argument for contract constructor.
  const parsedFee = ethers.utils.formatEther(deploymentFee.toString());

>   console.log(`The deployment will cost ${parsedFee} ETH`);

Reviewer: ```suggestion
  console.log(`The deployment is estimated to cost ${parsedFee} ETH`);
```

### Updating the greeting

> 1. Interacting with a smart contract works absolutely the same way as in `ethers`, however, if we want to use zkSync-specific features we may need to provide some additional parameters in the overrides:

Reviewer: We -> you? Below "you" is used, and I think it's a bit more expected form.

},
```

You now have a fully functioal Greeter-dApp! However, it does not leverage any zkSync-specific features.

### Paying fees using testnet paymaster

> Even though ether is the only token you can pay the fees with, the account abstraction feature allows you to integrate [paymasters](../zksync-v2/aa.md#paymasters) that can either pay the fees entirely for you or swap your tokens on the fly. In this tutorial, we will use the [testnet paymaster](../zksync-v2/aa.md#testnet-paymaster) that is provided on all zkSync testnets. It allows to pay fees in any ERC20 tokens with the exchange rate of ETH of 1:1, i.e. one wei of the token for one wei of ETH.

Author: I can change to "unit"

> Even though ether is the only token you can pay the fees with, the account abstraction feature allows you to integrate [paymasters](../zksync-v2/aa.md#paymasters) that can either pay the fees entirely for you or swap your tokens on the fly. In this tutorial, we will use the [testnet paymaster](../zksync-v2/aa.md#testnet-paymaster) that is provided on all zkSync testnets. It allows to pay fees in any ERC20 tokens with the exchange rate of ETH of 1:1, i.e. one wei of the token for one wei of ETH.

Reviewer 2: I don't mind, but that is why I added the " "

::: tip Mainnet integration

> Testnet paymaster is purely for demonstration of the feature and won't be available on mainnet. When integrating your protocol on mainnet, you should follow the documentation of the paymaster you will use.

Author: Hmm, it is not a "system" feature on testnet either. We just deployed a paymaster on our own and we are maintaining it. There are no patches to the system that made it available

> Testnet paymaster is purely for demonstration of the feature and won't be available on mainnet. When integrating your protocol on mainnet, you should follow the documentation of the paymaster you will use.

Reviewer 2: hmmm..... then "On mainnet, a paymaster won't be supported by the zkSync/ML team"?

> Testnet paymaster is purely for demonstration of the feature and won't be available on mainnet. When integrating your protocol on mainnet, you should follow the documentation of the paymaster you will use.

Author: > Testnet paymaster is purely for demonstration of the feature and won't be available on mainnet

Here we explicitly say that only the testnet paymaster won't be supported on zkSync 2.0

It is the same as if we decided for some reason to deploy our own AMM on testnet. We could do so on mainnet as well, but we won't

:::

> The address of the paymaster as well as the required input should be provided in the `getOverrides` method.

Author: Here are talking about return values and not inputs, so I think `in` is better here

> The address of the paymaster as well as the required input should be provided in the `getOverrides` method.

Reviewer 2: but you provide (the values) _to_ the method.

> 1. We need to retrieve the testnet paymaster:

Reviewer: ```suggestion
1. We need to retrieve the address of the testnet paymaster:
```

```javascript
async getOverrides() {
  if (this.selectedToken.l1Address != ETH_L1_ADDRESS) {
    const testnetPaymaster = await this.provider.getTestnetPaymasterAddress();

    // ..
  }

  return {};
}
```

Note, that it is recommended to retrieve the testnet paymaster's address each time before any interactions as it may change.

2. Add `utils` to the imports from `zksync-web3` SDK:

```javascript
import { Contract, Web3Provider, Provider, utils } from "zksync-web3";
```

> 2. We need to calculate how many tokens are required to process the transaction. Since the testnet paymaster exchanges any ERC20 token to ETH at 1:1 rate, then the number is the same as of ETH:

Reviewer 2: ```suggestion
2. We need to calculate how many tokens are required to process the transaction. Since the testnet paymaster exchanges any ERC20 token to ETH at a 1:1 rate, the number is the same as the ETH value:

Author: "ETH value" is a bad phrasing, since there is also `value` field which denotes the number of ETH passed with the transaction (not the fee)

Reviewer 2: edited
